test/performance/odp_ml_perf.c

```diff
@@ -0,0 +1,1051 @@
+			odp_ml_fp32_from_uint8((float *)out_d_addr, (uint8_t *)out_q_addr, elems,
+					       scale_d, 0);
+			break;
+		case ODP_ML_DATA_TYPE_FP16:
+			odp_ml_fp32_from_fp16((float *)out_d_addr, (uint16_t *)out_q_addr, elems);
+			break;
+		default:
+			ODPH_ERR("Unsupported type %d for output %d\n", info->data_type, i);
+			return -1;
+		}
+
+		out_q_addr += glb->out[i].size;
+		out_d_addr += elems * sizeof(float);
+	}
+
+	return 0;
+}
+
+static int test_ml(void *ptr)
+{
+	odp_time_t time;
+	odp_ml_data_seg_t inp_seg[MAX_IO];
+	uint8_t *inp_addr;
+	odp_ml_data_seg_t out_seg[MAX_IO];
+	uint8_t *out_addr;
+	int thread_idx = (int)(uintptr_t)ptr;
+	void *input = NULL, *output = NULL, *output_file = NULL;
+	void *output_final = NULL;
+	uint64_t out_size_final = 0;
+	int ret = 0;
+
+	input = malloc(glb->inp_size_q);
+
+	if (!input) {
+		ODPH_ERR("Allocating %" PRIu64 " bytes failed\n", glb->inp_size_q);
+		ret = -1;
+		goto error;
+	}
+
+	memcpy(input, glb->inp_file_data, glb->inp_size_q);
+	inp_addr = input;
+
+	if (glb->opt.scale_q > 0.0) {
+		input = malloc(glb->inp_size_q);
+		if (!input) {
+			ODPH_ERR("Allocating %" PRIu64 " bytes failed\n", glb->inp_size_q);
+			ret = -1;
+			goto error;
+		}
+
+		inp_addr = input;
+	}
+
+	for (int i = 0; i < glb->num_inp; i++) {
+		inp_seg[i].addr = inp_addr;
+		inp_seg[i].size = glb->inp[i].size;
+		inp_addr += glb->inp[i].size;
+	}
+
+	output = malloc(glb->out_size_q);
+
+	if (!output) {
+		ODPH_ERR("Allocating %" PRIu64 " bytes failed\n", glb->out_size_q);
+		ret = -1;
+		goto error;
+	}
+
+	out_addr = output;
+	output_final = output;
+	out_size_final = glb->out_size_q;
+
+	if (glb->opt.scale_d > 0.0) {
+		output_file = malloc(glb->out_size_d);
+		if (!output_file) {
+			ODPH_ERR("Allocating %" PRIu64 " bytes failed\n", glb->out_size_d);
+			ret = -1;
+			goto error;
+		}
+
+		output_final = output_file;
+		out_size_final = glb->out_size_d;
+	}
+
+	for (int i = 0; i < glb->num_out; i++) {
+		out_seg[i].addr = out_addr;
+		out_seg[i].size = glb->out[i].size;
+		out_addr += glb->out[i].size;
+	}
+
+	odp_ml_data_t data = {
+		.input_seg = inp_seg,
+		.num_input_seg = glb->num_inp,
+		.output_seg = out_seg,
+		.num_output_seg = glb->num_out,
+	};
+	odp_ml_run_param_t run_param;
+
+	odp_ml_run_param_init(&run_param);
+	run_param.batch_size = glb->opt.num_batch;
+
+	if (glb->opt.mode != MODE_INFERENCE_QUANT && glb->opt.scale_q > 0.0) {
+		if (quantize_input(input, glb->inp_file_data)) {
+			ret = -1;
+			goto error;
+		}
+	}
+
+	odp_barrier_wait(&glb->barrier);
+
+	for (int i = 0; i < glb->opt.rounds + glb->opt.warmup; i++) {
+		if (i == glb->opt.warmup)
+			time_start(&time);
+
+		if (glb->opt.mode == MODE_INFERENCE_QUANT && glb->opt.scale_q > 0.0) {
+			if (quantize_input(input, glb->inp_file_data)) {
+				ret = -1;
+				goto error;
+			}
+		}
+
+		int r;
+
+		while (!(r = odp_ml_run(glb->mdl, &data, &run_param)))
+			;
+		if (r != 1) {
+			ODPH_ERR("odp_ml_run() failed\n");
+			ret = -1;
+			goto error;
+		}
```

Comment:
r == 0 may be sporadic, other errors probably not. Either way, if anything went wrong, there's no point in continuing the test because what's the value of a performace test result if there was an error during the test. It's better to just quit and exit with an error status.